<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#HOME-CREDIT-MODELING-ASSIGNMENT-GROUP-2---Justin-Hamilton-Contribution" data-toc-modified-id="HOME-CREDIT-MODELING-ASSIGNMENT-GROUP-2---Justin-Hamilton-Contribution-1">HOME CREDIT MODELING ASSIGNMENT GROUP 2 - Justin Hamilton Contribution</a></span></li><li><span><a href="#Data-Preparation" data-toc-modified-id="Data-Preparation-2">Data Preparation</a></span><ul class="toc-item"><li><span><a href="#Importing-Necessary-Packages" data-toc-modified-id="Importing-Necessary-Packages-2.1">Importing Necessary Packages</a></span></li><li><span><a href="#Making-the-dataframes-for-each-of-our-CSV-files" data-toc-modified-id="Making-the-dataframes-for-each-of-our-CSV-files-2.2">Making the dataframes for each of our CSV files</a></span></li><li><span><a href="#Combining-the-DFs-for-our-Analysis" data-toc-modified-id="Combining-the-DFs-for-our-Analysis-2.3">Combining the DFs for our Analysis</a></span></li><li><span><a href="#Multiple-Regression-Models" data-toc-modified-id="Multiple-Regression-Models-2.4">Multiple Regression Models</a></span><ul class="toc-item"><li><span><a href="#Fitting-several-different-logistic-regression-models-using-different-interactions" data-toc-modified-id="Fitting-several-different-logistic-regression-models-using-different-interactions-2.4.1">Fitting several different logistic regression models using different interactions</a></span></li><li><span><a href="#Attempting-for-the-Up-and-Down-Sampling-for-the-Logistic-Regression-Model" data-toc-modified-id="Attempting-for-the-Up-and-Down-Sampling-for-the-Logistic-Regression-Model-2.4.2">Attempting for the Up and Down Sampling for the Logistic Regression Model</a></span></li><li><span><a href="#Hyper-Parameter-Tuning-of-the-Logistic-Regressions" data-toc-modified-id="Hyper-Parameter-Tuning-of-the-Logistic-Regressions-2.4.3">Hyper Parameter Tuning of the Logistic Regressions</a></span></li><li><span><a href="#Model-performance-of-Best-Logistic-Regression-Model" data-toc-modified-id="Model-performance-of-Best-Logistic-Regression-Model-2.4.4">Model performance of Best Logistic Regression Model</a></span></li><li><span><a href="#Features-of-the-Best-Logistic-Model" data-toc-modified-id="Features-of-the-Best-Logistic-Model-2.4.5">Features of the Best Logistic Model</a></span></li><li><span><a href="#Prediction-on-the-Test-Set" data-toc-modified-id="Prediction-on-the-Test-Set-2.4.6">Prediction on the Test Set</a></span></li></ul></li><li><span><a href="#Random-Forest-Classification-Model" data-toc-modified-id="Random-Forest-Classification-Model-2.5">Random Forest Classification Model</a></span><ul class="toc-item"><li><span><a href="#Creating-the-Random-Forest-Classification-and-printing-the-Accuracy-and-AUC-of-the-Model" data-toc-modified-id="Creating-the-Random-Forest-Classification-and-printing-the-Accuracy-and-AUC-of-the-Model-2.5.1">Creating the Random Forest Classification and printing the Accuracy and AUC of the Model</a></span></li><li><span><a href="#Attempting-for-the-Up-and-Down-Sampling-for-the-Random-Forest-Model" data-toc-modified-id="Attempting-for-the-Up-and-Down-Sampling-for-the-Random-Forest-Model-2.5.2">Attempting for the Up and Down Sampling for the Random Forest Model</a></span></li><li><span><a href="#Hyper-Parameters" data-toc-modified-id="Hyper-Parameters-2.5.3">Hyper Parameters</a></span></li><li><span><a href="#Evaluating-the-Best-Model-Performance-of-the-Random-Classification" data-toc-modified-id="Evaluating-the-Best-Model-Performance-of-the-Random-Classification-2.5.4">Evaluating the Best Model Performance of the Random Classification</a></span></li><li><span><a href="#Features-of-the-Best-Random-Forest-Classification-Model" data-toc-modified-id="Features-of-the-Best-Random-Forest-Classification-Model-2.5.5">Features of the Best Random Forest Classification Model</a></span></li><li><span><a href="#Prediction-on-the-Test-Set" data-toc-modified-id="Prediction-on-the-Test-Set-2.5.6">Prediction on the Test Set</a></span></li></ul></li><li><span><a href="#Ensemble-Model" data-toc-modified-id="Ensemble-Model-2.6">Ensemble Model</a></span><ul class="toc-item"><li><span><a href="#Creating-the-Ensemble-Model-with-the-RandomForest,-Logistic-Regression,-and-Gradient-Boosting-Models" data-toc-modified-id="Creating-the-Ensemble-Model-with-the-RandomForest,-Logistic-Regression,-and-Gradient-Boosting-Models-2.6.1">Creating the Ensemble Model with the RandomForest, Logistic Regression, and Gradient Boosting Models</a></span></li><li><span><a href="#Up-and-Down-Sampling" data-toc-modified-id="Up-and-Down-Sampling-2.6.2">Up and Down Sampling</a></span></li><li><span><a href="#Hyper-Parameters" data-toc-modified-id="Hyper-Parameters-2.6.3">Hyper Parameters</a></span></li><li><span><a href="#Evaluating-the-Best-of-the-Ensemble-Model" data-toc-modified-id="Evaluating-the-Best-of-the-Ensemble-Model-2.6.4">Evaluating the Best of the Ensemble Model</a></span></li><li><span><a href="#Features-of-the-Best-Ensemble-Model" data-toc-modified-id="Features-of-the-Best-Ensemble-Model-2.6.5">Features of the Best Ensemble Model</a></span></li><li><span><a href="#Prediction-on-the-Test-Set" data-toc-modified-id="Prediction-on-the-Test-Set-2.6.6">Prediction on the Test Set</a></span></li></ul></li></ul></li><li><span><a href="#Model-Performances-Summary" data-toc-modified-id="Model-Performances-Summary-3">Model Performances Summary</a></span></li><li><span><a href="#Results-Summary" data-toc-modified-id="Results-Summary-4">Results Summary</a></span></li><li><span><a href="#Justin-Hamilton-Individual-Contribution-Summary" data-toc-modified-id="Justin-Hamilton-Individual-Contribution-Summary-5">Justin Hamilton Individual Contribution Summary</a></span></li></ul></div>

# HOME CREDIT MODELING ASSIGNMENT GROUP 2 - Justin Hamilton Contribution

# Data Preparation

## Importing Necessary Packages

In [1]:
import importlib

# List of required packages
required_packages = [
    'pandas',
    'numpy',
    'matplotlib',
    'xgboost',
    'imbalanced-learn',
    'scipy',
    'sklearn',
]

# Check and install missing packages
for package in required_packages:
    try:
        importlib.import_module(package)
        print(f"{package} is already installed.")
    except ImportError:
        print(f"{package} is not installed. Installing...")
        !pip install {package}

# Import the required modules
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import xgboost as xgb

from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_auc_score
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, roc_auc_score
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder
import numpy as np

from sklearn.utils import resample
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
import pandas as pd

from sklearn.ensemble import VotingClassifier
from sklearn.model_selection import cross_val_score

from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import classification_report, roc_auc_score

# from imblearn.over_sampling import RandomOverSampler
from scipy.stats import randint, uniform
from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV, cross_val_score
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.metrics import accuracy_score, classification_report, roc_auc_score, confusion_matrix
from sklearn.ensemble import RandomForestClassifier

# Additional configurations or settings
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline



pandas is already installed.
numpy is already installed.
matplotlib is already installed.
xgboost is already installed.
imbalanced-learn is not installed. Installing...
sklearn is already installed.



Many of these packages have already been installed. Each of these packages serves different purposes within our code for each of the modeling. All of the packages have imported correctly and have succesfully been installed for use.

## Making the dataframes for each of our CSV files

In [2]:
# Making Dataframes of each of the CSV Files

applicationtestDF = pd.read_csv("application_test.csv")
applicationtrainDF = pd.read_csv("application_train.csv")
bureauDF = pd.read_csv("bureau.csv")
bureaubalanceDF = pd.read_csv("bureau_balance.csv")
creditcardbalanceDF = pd.read_csv("credit_card_balance.csv")
installmentpaymentsDF = pd.read_csv("installments_payments.csv")
poscashDF = pd.read_csv("POS_CASH_balance.csv")
previousDF = pd.read_csv("previous_application.csv")

In [3]:
#Shape of datasets
print(applicationtestDF.shape)
print(applicationtrainDF.shape)
print(bureauDF.shape)
print(bureaubalanceDF.shape)
print(creditcardbalanceDF.shape)
print(installmentpaymentsDF.shape)
print(poscashDF.shape)
print(previousDF.shape)

(48744, 121)
(307511, 122)
(1716428, 17)
(27299925, 3)
(3840312, 23)
(13605401, 8)
(10001358, 8)
(1670214, 37)


Each of the CSV were loaded and installed successfully. Due to the size of each of the CSV files, we have to be selected as to what files to use because it would require too much computation power to use all. Many of the CSV files have data points that are shared in common with other files. We will go into more analysis of which ones to combine into our aggregation of the data sets in our future analysis

## Combining the DFs for our Analysis

***The original approach that we took to getting the DF into one combined Dataframe to analyze as best as possible was to take this approach***

In [4]:
# # # Combine the DataFrames
# combined_df = pd.concat([applicationtestDF, applicationtrainDF, bureauDF, bureaubalanceDF,
#                          creditcardbalanceDF, installmentpaymentsDF, poscashDF, previousDF])


# Remove missing values (NA)
# combined_df = combined_df.dropna()

# # Remove duplicates from the combined DataFrame
# combined_df = combined_df.drop_duplicates()

# # Reset the index
# combined_df = combined_df.reset_index(drop=True)

# # Print the head of the combined DataFrame
# combined_df.head()


# # Print the data types of each column
# combined_df.dtypes




However, this approach took up too much computational power to execute due to the size of the data. The append() function was then tried to be used other than the concat() to save memory. This execution was not very appropriate because each of our devices did not have the computation power at first to just combine all datasets into one. We then had decided to continue on and try to use CSV files that we felt are only relevant to our analysis

In [5]:
# combined_df = pd.DataFrame()  # Initialize an empty DataFrame

# # List of DataFrames to be combined
# dataframes = [applicationtestDF, applicationtrainDF, bureauDF, bureaubalanceDF,
#               creditcardbalanceDF, installmentpaymentsDF, poscashDF, previousDF]

# # Combine DataFrames in smaller batches
# for df in dataframes:
#     combined_df = combined_df.append(df, ignore_index=True)

# # Remove missing values (NA)
# combined_df = combined_df.dropna()

# # Remove duplicates from the combined DataFrame
# combined_df = combined_df.drop_duplicates()

# # Reset the index
# combined_df = combined_df.reset_index(drop=True)

# # Print the head of the combined DataFrame
# combined_df.head()

# # Print the data types of each column
# combined_df.dtypes


This method did not work. So We then aggregated some of the values based upon the averages of each datasets. We thought the count of days since overdue payments on both credit card balance and the bureau tables would be good to examine. This method gave us a much better output. By using merge() rather than concat() we are able to merge datasets that fit within computation and add elements to understand for our modeling. We have chosen the creditcardbalance and bureauDF because these dataframes contain much of the same info from the other DF. This information also can tell us how customer behavior may lead to default.

In [6]:
#Calculate average 'SK_DPD' values in credit_card_balance.csv
average_sk_dpd = creditcardbalanceDF.groupby('SK_ID_CURR')['SK_DPD'].mean().reset_index()

#Merge average_sk_dpd with train_data based on 'SK_ID_CURR'
train_data = applicationtrainDF.merge(average_sk_dpd, on='SK_ID_CURR', how='left')

#Merge average_sk_dpd with test_data based on 'SK_ID_CURR'
test_data = applicationtestDF.merge(average_sk_dpd, on='SK_ID_CURR', how='left')

#Fill missing values with 0
train_data['SK_DPD'] = train_data['SK_DPD'].fillna(0)
test_data['SK_DPD'] = test_data['SK_DPD'].fillna(0)

#Calculate average 'CREDIT_DAY_OVERDUE' values in bureau.csv
average_credit_day_overdue = bureauDF.groupby('SK_ID_CURR')['CREDIT_DAY_OVERDUE'].mean().reset_index()

#Merge average_credit_day_overdue with train_data based on 'SK_ID_CURR'
train_data = train_data.merge(average_credit_day_overdue, on='SK_ID_CURR', how='left')

#Merge average_credit_day_overdue with test_data based on 'SK_ID_CURR'
test_data = test_data.merge(average_credit_day_overdue, on='SK_ID_CURR', how='left')

#Fill missing values with 0
train_data['CREDIT_DAY_OVERDUE'] = train_data['CREDIT_DAY_OVERDUE'].fillna(0)
test_data['CREDIT_DAY_OVERDUE'] = test_data['CREDIT_DAY_OVERDUE'].fillna(0)

#Shape of the joined datasets
print(train_data.shape)
print(test_data.shape)

#First 10 rows on the joined dataset
train_data.head(10)

(307511, 124)
(48744, 123)


SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   
5      100008       0         Cash loans           M            N   
6      100009       0         Cash loans           F            Y   
7      100010       0         Cash loans           M            Y   
8      100011       0         Cash loans           F            N   
9      100012       0    Revolving loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   
5               Y             0           99000.0    490495.5      27517.5   
6               Y             1          171000.0   1560726.0      41301.0   
7               Y             0          360000.0   1530000.0      42075.0   
8               Y             0          112500.0   1019610.0      33826.5   
9               Y             0          135000.0    405000.0      20250.0   

   ...  FLAG_DOCUMENT_20 FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR  \
0  ...                 0                0                        0.0   
1  ...                 0                0                        0.0   
2  ...                 0                0                        0.0   
3  ...                 0                0                        NaN   
4  ...                 0                0                        0.0   
5  ...                 0                0                        0.0   
6  ...                 0                0                        0.0   
7  ...                 0                0                        0.0   
8  ...                 0                0                        0.0   
9  ...                 0                0                        NaN   

  AMT_REQ_CREDIT_BUREAU_DAY AMT_REQ_CREDIT_BUREAU_WEEK  \
0                       0.0                        0.0   
1                       0.0                        0.0   
2                       0.0                        0.0   
3                       NaN                        NaN   
4                       0.0                        0.0   
5                       0.0                        0.0   
6                       0.0                        0.0   
7                       0.0                        0.0   
8                       0.0                        0.0   
9                       NaN                        NaN   

  AMT_REQ_CREDIT_BUREAU_MON  AMT_REQ_CREDIT_BUREAU_QRT  \
0                       0.0                        0.0   
1                       0.0                        0.0   
2                       0.0                        0.0   
3                       NaN                        NaN   
4                       0.0                        0.0   
5                       0.0                        1.0   
6                       1.0                        1.0   
7                       0.0                        0.0   
8                       0.0                        0.0   
9                       NaN                        NaN   

   AMT_REQ_CREDIT_BUREAU_YEAR  SK_DPD  CREDIT_DAY_OVERDUE  
0                         1.0     0.0                 0.0  
1                         0.0     0.0                 0.0  
2                         0.0     0.0                 0.0  
3                         NaN     0.0                 0.0  
4            

This method worked a lot better to get our dataframes merged to set aside for our testing and training datasets.

## Multiple Regression Models

### Fitting several different logistic regression models using different interactions

In [61]:
import time
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_auc_score

# Generate interaction terms
poly = PolynomialFeatures(interaction_only=True, include_bias=False)
start_time = time.time()
X_train_with_interactions = poly.fit_transform(X_train)
end_time = time.time()

# Fit logistic regression models with and without interaction terms
model_without_interactions = LogisticRegression()
model_without_interactions.fit(X_train, y_train)

model_with_interactions = LogisticRegression()
start_time_with_interactions = time.time()
model_with_interactions.fit(X_train_with_interactions, y_train)
end_time_with_interactions = time.time()

# Prepare validation data
start_time_val = time.time()
X_val_with_interactions = poly.transform(X_val)
end_time_val = time.time()

# Make predictions on the validation set
start_time_preds = time.time()
preds_without_interactions = model_without_interactions.predict(X_val)
probs_without_interactions = model_without_interactions.predict_proba(X_val)[:, 1]

preds_with_interactions = model_with_interactions.predict(X_val_with_interactions)
probs_with_interactions = model_with_interactions.predict_proba(X_val_with_interactions)[:, 1]
end_time_preds = time.time()

# Evaluate model performance
start_time_eval = time.time()
accuracy_without_interactions = accuracy_score(y_val, preds_without_interactions)
auc_without_interactions = roc_auc_score(y_val, probs_without_interactions)

accuracy_with_interactions = accuracy_score(y_val, preds_with_interactions)
auc_with_interactions = roc_auc_score(y_val, probs_with_interactions)
end_time_eval = time.time()

# Print the runtimes
print("Runtime for generating interaction terms:", end_time - start_time, "seconds")
print("Runtime for fitting model with interactions:", end_time_with_interactions - start_time_with_interactions, "seconds")
print("Runtime for preparing validation data:", end_time_val - start_time_val, "seconds")
print("Runtime for making predictions:", end_time_preds - start_time_preds, "seconds")
print("Runtime for evaluating model performance:", end_time_eval - start_time_eval, "seconds")

# Compare model performance
print("\nModel without interactions:")
print("Accuracy:", accuracy_without_interactions)
print("AUC:", auc_without_interactions)

print("\nModel with interactions:")
print("Accuracy:", accuracy_with_interactions)
print("AUC:", auc_with_interactions)


Runtime for generating interaction terms: 11.300916910171509 seconds
Runtime for fitting model with interactions: 18.78170418739319 seconds
Runtime for preparing validation data: 2.673457384109497 seconds
Runtime for making predictions: 1.056178331375122 seconds
Runtime for evaluating model performance: 0.1746351718902588 seconds

Model without interactions:
Accuracy: 0.9195161211648212
AUC: 0.6203350955928995

Model with interactions:
Accuracy: 0.9195323805342829
AUC: 0.5


From our output of the logistic regression modeling, we see that the model without interactions had a accuracy of 91.9516% and an AUC of .62. The model with interactions has a accuracy of 91.9532% but an AUC of .5.


There may be a higher AUC within the model without interactions because there is less to consider in the effect that interactions can have one the target variable. Although our AUC is lower for the model with interactions, this may because now we are looking at the bigger picture with all interactive variables and we are resulting in getting a higher accuracy of the logistic regression model when we include the interactions. Not as high of an AUC as it should be because this may signal overfitting or random guessing.

### Attempting for the Up and Down Sampling for the Logistic Regression Model

In [62]:
import time
from sklearn.linear_model import LogisticRegression
from sklearn.utils import resample
from sklearn.metrics import accuracy_score

# Analyze class distribution in the training set
class_counts = y_train.value_counts()
print("Class Distribution (Before):")
print(class_counts)

# Upsample the minority class
df_majority = X_train[y_train == 0]
df_minority = X_train[y_train == 1]

start_time_upsample = time.time()
df_minority_upsampled = resample(df_minority, replace=True, n_samples=len(df_majority), random_state=42)
X_train_upsampled = pd.concat([df_majority, df_minority_upsampled])
y_train_upsampled = pd.Series([0] * len(df_majority) + [1] * len(df_minority_upsampled))
end_time_upsample = time.time()

# Downsample the majority class
start_time_downsample = time.time()
df_majority_downsampled = resample(df_majority, replace=False, n_samples=len(df_minority), random_state=42)
X_train_downsampled = pd.concat([df_majority_downsampled, df_minority])
y_train_downsampled = pd.Series([0] * len(df_majority_downsampled) + [1] * len(df_minority))
end_time_downsample = time.time()

# Train and evaluate models on upsampled and downsampled data
model = LogisticRegression()  # Initialize the LogisticRegression model

# Train on upsampled data
start_time_train_upsampled = time.time()
model.fit(X_train_upsampled, y_train_upsampled)
y_pred_upsampled = model.predict(X_val)
accuracy_upsampled = accuracy_score(y_val, y_pred_upsampled)
end_time_train_upsampled = time.time()

# Train on downsampled data
start_time_train_downsampled = time.time()
model.fit(X_train_downsampled, y_train_downsampled)
y_pred_downsampled = model.predict(X_val)
accuracy_downsampled = accuracy_score(y_val, y_pred_downsampled)
end_time_train_downsampled = time.time()

# Compute predictions for original data
start_time_pred_original = time.time()
y_pred_original = model.predict(X_val)
accuracy_original = accuracy_score(y_val, y_pred_original)
end_time_pred_original = time.time()

# Print the runtimes
print("Upsampling Runtime:", end_time_upsample - start_time_upsample, "seconds")
print("Downsampling Runtime:", end_time_downsample - start_time_downsample, "seconds")
print("Upsampled Training Runtime:", end_time_train_upsampled - start_time_train_upsampled, "seconds")
print("Downsampled Training Runtime:", end_time_train_downsampled - start_time_train_downsampled, "seconds")
print("Original Prediction Runtime:", end_time_pred_original - start_time_pred_original, "seconds")

# Compare model performance
print("Accuracy (Original):", accuracy_original)
print("Accuracy (Upsampled):", accuracy_upsampled)
print("Accuracy (Downsampled):", accuracy_downsampled)


Class Distribution (Before):
0    226132
1     19876
Name: TARGET, dtype: int64
Upsampling Runtime: 1.366769790649414 seconds
Downsampling Runtime: 0.23794841766357422 seconds
Upsampled Training Runtime: 7.301072359085083 seconds
Downsampled Training Runtime: 0.6300501823425293 seconds
Original Prediction Runtime: 0.03173041343688965 seconds
Accuracy (Original): 0.5965237468090987
Accuracy (Upsampled): 0.5941986569760824
Accuracy (Downsampled): 0.5965237468090987


When considering whether to up or downsample the dataset, we see that upsampling gives us the lowest accuracy within our modeling of only being 59.41% accurate in predicting for the target variable.

We see that the original and downsample of the logistic regression give us the same and higher accuracy of 59.65%. Because of this when we perform our hyper parameters and evaluations within the logistic regression model we are going to keep our original data rather than the up or down sampled versions of the data.

### Hyper Parameter Tuning of the Logistic Regressions

In [63]:
import time
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Define the hyperparameters and their search space
param_grid = {
    'C': [0.1, 1.0, 10.0],
    'penalty': ['l1', 'l2'],
    'solver': ['liblinear']
}

# Initialize the model
model = LogisticRegression(random_state=42)

# Perform grid search with cross-validation
start_time = time.time()
grid_search = GridSearchCV(model, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, y_train)
end_time = time.time()

# Get the best hyperparameter values and model performance
best_params = grid_search.best_params_
best_score = grid_search.best_score_

# Train the model with the best hyperparameters
start_time_train = time.time()
best_model = LogisticRegression(**best_params, random_state=42)
best_model.fit(X_train, y_train)
end_time_train = time.time()

# Make predictions using the best model
start_time_pred = time.time()
y_pred = best_model.predict(X_val)
end_time_pred = time.time()

# Evaluate the model performance
start_time_eval = time.time()
accuracy = accuracy_score(y_val, y_pred)
end_time_eval = time.time()

# Print the runtimes
print("Grid Search Runtime:", end_time - start_time, "seconds")
print("Training Runtime:", end_time_train - start_time_train, "seconds")
print("Prediction Runtime:", end_time_pred - start_time_pred, "seconds")
print("Evaluation Runtime:", end_time_eval - start_time_eval, "seconds")

# Print the best hyperparameters and model performance
print("Best Hyperparameters:", best_params)
print("Best Score:", best_score)
print("Accuracy:", accuracy)


Grid Search Runtime: 364.85560846328735 seconds
Training Runtime: 4.806471347808838 seconds
Prediction Runtime: 0.051279544830322266 seconds
Evaluation Runtime: 0.0 seconds
Best Hyperparameters: {'C': 0.1, 'penalty': 'l2', 'solver': 'liblinear'}
Best Score: 0.9192018146332195
Accuracy: 0.9195161211648212


After using hyper parameters, we see a couple of interesting observations. The first of which is that this process takes a long time to complete. Taking about 6 minutes to compute.

Within the best hyperparameters for modeling we see that

* Because C is .1, suggests a moderately strong regularization is preferred by the model
* The penalty being of L2 means that Ridge Regularization will give us our best results
* Solver being assigned as liblinear means that this solver is suitable for small to medium size datasets and supports L1 and L2 penalities


With the best score being 0.9192018146332195 this means that at the cross validations the mean of the accuracies is 0.9192018146332195. Meaning, that on average at the best parameters the model was performing at a accuracy that is a little too high. This may result because of over fitting within the model performed by the machine. 

### Model performance of Best Logistic Regression Model

In [64]:
import time
from sklearn.metrics import classification_report, roc_auc_score

# Train set performance
start_time_train = time.time()
train_pred = best_model.predict(X_train)
train_auc = roc_auc_score(y_train, train_pred)
end_time_train = time.time()

# Print train set performance and runtime
print("Train Set Performance:")
print("AUC:", train_auc)
print(classification_report(y_train, train_pred))
print("Train Set Runtime:", end_time_train - start_time_train, "seconds")

# Test set performance
start_time_test = time.time()
test_pred = best_model.predict(X_val)
test_auc = roc_auc_score(y_val, test_pred)
end_time_test = time.time()

# Print test set performance and runtime
print("Test Set Performance:")
print("AUC:", test_auc)
print(classification_report(y_val, test_pred))
print("Test Set Runtime:", end_time_test - start_time_test, "seconds")


Train Set Performance:
AUC: 0.4999977889020572
              precision    recall  f1-score   support

           0       0.92      1.00      0.96    226132
           1       0.00      0.00      0.00     19876

    accuracy                           0.92    246008
   macro avg       0.46      0.50      0.48    246008
weighted avg       0.84      0.92      0.88    246008

Train Set Runtime: 0.22581863403320312 seconds
Test Set Performance:
AUC: 0.49999115889238605
              precision    recall  f1-score   support

           0       0.92      1.00      0.96     56554
           1       0.00      0.00      0.00      4949

    accuracy                           0.92     61503
   macro avg       0.46      0.50      0.48     61503
weighted avg       0.85      0.92      0.88     61503

Test Set Runtime: 0.03416752815246582 seconds


Under the best parameters set for the Logistic Regression Modeling

The Train Set
* Has an AUC of 0.4999977889020572. Meaning that the models predictive powers seem to be pretty close to random and the classification power of the model are fairly low.
* The precision, recall, and F1 score for the majority class of 0 (Those who do not default) are all very high meaning that the model performs well in identifying the majority class
* However the precision, recall, and F1 score for the minority class of 1 (Those who do default) are very low at 0. Meaning that it is difficult for the model to predict those who will default.
* This train set does give a 92% accuracy though, largely due to its ability to correctly predict the majority class of those who do not default

The Test Set

* The evaluation of the Test set performs almost identical to that of the train dataset.


This shows us that the Logistic regression model is a model that does accurately and have a power to predict those who are safe to borrow to, which is the purpose of the model. 

While the AUC may be lower and may suggest low discrimative power in classification, this may be because of overfitting performed within the parameters of the model. 



### Features of the Best Logistic Model

In [65]:
import time
from sklearn.linear_model import LogisticRegression
import pandas as pd

# Set pandas display options
pd.set_option('display.float_format', lambda x: '%.8f' % x)

# Get the feature coefficients and their corresponding names
start_time = time.time()
coefficients = best_model.coef_[0]
feature_names = X_train.columns
end_time = time.time()

# Create a DataFrame to store the coefficients and feature names
start_time_df = time.time()
coefficients_df = pd.DataFrame({'Feature': feature_names, 'Coefficient': coefficients})
end_time_df = time.time()

# Sort the DataFrame by coefficient magnitude in descending order
start_time_sort = time.time()
coefficients_df = coefficients_df.sort_values('Coefficient', ascending=False)
end_time_sort = time.time()

# Display the coefficients and their impact on the target variable
print(coefficients_df.head(10))

# Print the runtimes
print("Runtime for retrieving coefficients:", end_time - start_time, "seconds")
print("Runtime for creating DataFrame:", end_time_df - start_time_df, "seconds")
print("Runtime for sorting DataFrame:", end_time_sort - start_time_sort, "seconds")


                       Feature  Coefficient
33      DAYS_LAST_PHONE_CHANGE   0.00021637
10             DAYS_ID_PUBLISH   0.00009348
7                   DAYS_BIRTH   0.00008404
9            DAYS_REGISTRATION   0.00002685
4                  AMT_ANNUITY   0.00000776
3                   AMT_CREDIT   0.00000310
61          CREDIT_DAY_OVERDUE   0.00000259
8                DAYS_EMPLOYED   0.00000052
59  AMT_REQ_CREDIT_BUREAU_YEAR   0.00000048
60                      SK_DPD   0.00000025
Runtime for retrieving coefficients: 0.0 seconds
Runtime for creating DataFrame: 0.024721384048461914 seconds
Runtime for sorting DataFrame: 0.009205102920532227 seconds


From this output, we learn that when each of these things increases, so does the likelihood of the person being safe to borrow to increases

* Number of days since the clients last owned phone
* Number of days since the client changed their ID info
* Age in days of the client
* Number of days before the application the client change their registration
* Loan annuity
* Credit amount of the loan
* Number of days past due on CB credit at the time of application for related loan
* Days before the application the person started current employment
* Number of enquiries to Credit Bureau about the client one year
* DPD (Days past due) during the month on the previous credit

### Prediction on the Test Set

In [66]:
#Make predictions on the test set
test_predictions = best_model.predict(test_data)

In [68]:
#Create a submission DataFrame
submission = pd.DataFrame({
    "SK_ID_CURR": test_data["SK_ID_CURR"], 
    "TARGET": test_predictions
})

#Remove duplicate SK_ID_CURR values
submission = submission.drop_duplicates(subset="SK_ID_CURR", keep="first")

#Save the submission DataFrame to a CSV file
submission.to_csv("submission_log2.csv", index=False)

When we put this into Kaggle, we get a score of .5 While this may not be as high as anticipated, this does align with the AUC that signaled not as satisfactory results of potentially random guessing within the data set rather than making accurate predictions. 

## Random Forest Classification Model

### Creating the Random Forest Classification and printing the Accuracy and AUC of the Model

In [88]:
import time
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, roc_auc_score

# Fit Random Forest model
start_time_fit = time.time()
rf_model = RandomForestClassifier()
rf_model.fit(X_train, y_train)
end_time_fit = time.time()

# Make predictions with Random Forest model
start_time_pred = time.time()
rf_preds = rf_model.predict(X_val)
rf_probs = rf_model.predict_proba(X_val)[:, 1]
end_time_pred = time.time()

# Evaluate Random Forest model performance
start_time_eval = time.time()
rf_accuracy = accuracy_score(y_val, rf_preds)
rf_auc = roc_auc_score(y_val, rf_probs)
end_time_eval = time.time()

# Print the runtimes
print("Fit Runtime:", end_time_fit - start_time_fit, "seconds")
print("Prediction Runtime:", end_time_pred - start_time_pred, "seconds")
print("Evaluation Runtime:", end_time_eval - start_time_eval, "seconds")

# Compare model performance
print("Random Forest Model:")
print("Accuracy:", rf_accuracy)
print("AUC:", rf_auc)


Fit Runtime: 147.61922216415405 seconds
Prediction Runtime: 4.936388731002808 seconds
Evaluation Runtime: 0.02080988883972168 seconds
Random Forest Model:
Accuracy: 0.9196136773815912
AUC: 0.7069750901140923


When we perform a raw Random Forest Model we see that we get an accuracy of 0.919711233598361 or about 92%. This is a very high accuracy meaning that this model does not do a great job at predicting the accuracy of the majority class. We have an AUC of 0.7069800939416186 meaning that the Random Forest Model has a strong discrimative power to tell if a customer would be a safe borrower or not. 

###  Attempting for the Up and Down Sampling for the Random Forest Model

In [89]:
import time
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.utils import resample

# Analyze class distribution in the training set
start_time_dist = time.time()
class_counts = y_train.value_counts()
end_time_dist = time.time()

print("Class Distribution (Before):")
print(class_counts)

# Upsample the minority class
start_time_upsample = time.time()
df_majority = X_train[y_train == 0]
df_minority = X_train[y_train == 1]

df_minority_upsampled = resample(df_minority, replace=True, n_samples=len(df_majority), random_state=42)

X_train_upsampled = pd.concat([df_majority, df_minority_upsampled])
y_train_upsampled = pd.Series([0] * len(df_majority) + [1] * len(df_minority_upsampled))
end_time_upsample = time.time()

# Downsample the majority class
start_time_downsample = time.time()
df_majority_downsampled = resample(df_majority, replace=False, n_samples=len(df_minority), random_state=42)

X_train_downsampled = pd.concat([df_majority_downsampled, df_minority])
y_train_downsampled = pd.Series([0] * len(df_majority_downsampled) + [1] * len(df_minority))
end_time_downsample = time.time()

# Train and evaluate models on upsampled and downsampled data
start_time_train = time.time()
model = RandomForestClassifier()  # Initialize the Random Forest model

# Train on upsampled data
model.fit(X_train_upsampled, y_train_upsampled)
y_pred_upsampled = model.predict(X_val)
accuracy_upsampled = accuracy_score(y_val, y_pred_upsampled)

# Train on downsampled data
model.fit(X_train_downsampled, y_train_downsampled)
y_pred_downsampled = model.predict(X_val)
accuracy_downsampled = accuracy_score(y_val, y_pred_downsampled)
end_time_train = time.time()

# Compute predictions for original data
start_time_pred = time.time()
y_pred_original = model.predict(X_val)
accuracy_original = accuracy_score(y_val, y_pred_original)
end_time_pred = time.time()

# Compare model performance
print("Accuracy (Original):", accuracy_original)
print("Accuracy (Upsampled):", accuracy_upsampled)
print("Accuracy (Downsampled):", accuracy_downsampled)

# Print the runtimes
print("Class Distribution Runtime:", end_time_dist - start_time_dist, "seconds")
print("Upsampling Runtime:", end_time_upsample - start_time_upsample, "seconds")
print("Downsampling Runtime:", end_time_downsample - start_time_downsample, "seconds")
print("Training Runtime:", end_time_train - start_time_train, "seconds")
print("Prediction Runtime:", end_time_pred - start_time_pred, "seconds")


Class Distribution (Before):
0    226132
1     19876
Name: TARGET, dtype: int64
Accuracy (Original): 0.6793164561078321
Accuracy (Upsampled): 0.9196299367510528
Accuracy (Downsampled): 0.6793164561078321
Class Distribution Runtime: 0.013184785842895508 seconds
Upsampling Runtime: 0.5757849216461182 seconds
Downsampling Runtime: 0.07246971130371094 seconds
Training Runtime: 213.34045934677124 seconds
Prediction Runtime: 1.3933310508728027 seconds


When we test the accuracy of the data on the Random Forest Model, we see that it has the highest accuracy when the data is upsampled. Upsampmling gives us an accuracy of 0.9194998617953596 or 91.95%. While the original and downsampled version of the dataset gives us a moderate accuracy of 0.6786823406988277 or 67.87%. Although accuracy may not be as beneficial as using the measurement of AUC.

### Hyper Parameters

In [91]:
import time
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Define the hyperparameters and their search space
param_grid = {
    'n_estimators': [100, 200],
    'max_depth': [None, 5],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2]
}

# Initialize the model
model = RandomForestClassifier(random_state=42)

# Perform randomized search with cross-validation
start_time_search = time.time()
# CV 2 was set here rather than 5 because the CV 5 took too much time. Longer than 8 hours. n_iter = 5 
# was also set to make the process go faster
random_search = RandomizedSearchCV(model, param_distributions=param_grid, n_iter=5, cv=2, scoring='accuracy')
# Because our UpSampling had a better accuracy, we are using that here rather than the original 
random_search.fit(X_train_upsampled, y_train_upsampled)
end_time_search = time.time()

# Get the best hyperparameter values and model performance
best_params = random_search.best_params_
best_score = random_search.best_score_

# Train the model with the best hyperparameters
start_time_train = time.time()
best_model = RandomForestClassifier(**best_params, random_state=42)
best_model.fit(X_train_upsampled, y_train_upsampled)
end_time_train = time.time()

# Make predictions using the best model
start_time_pred = time.time()
y_pred = best_model.predict(X_val)
end_time_pred = time.time()

# Evaluate the model performance
start_time_eval = time.time()
accuracy = accuracy_score(y_val, y_pred)
end_time_eval = time.time()

# Print the best hyperparameters and model performance
print("Best Hyperparameters:", best_params)
print("Best Score:", best_score)
print("Accuracy:", accuracy)

# Print the runtimes
print("Randomized Search Runtime:", end_time_search - start_time_search, "seconds")
print("Training Runtime:", end_time_train - start_time_train, "seconds")
print("Prediction Runtime:", end_time_pred - start_time_pred, "seconds")
print("Evaluation Runtime:", end_time_eval - start_time_eval, "seconds")


Best Hyperparameters: {'n_estimators': 100, 'min_samples_split': 2, 'min_samples_leaf': 1, 'max_depth': None}
Best Score: 0.9900522703553676
Accuracy: 0.9194185649480513
Randomized Search Runtime: 652.6474483013153 seconds
Training Runtime: 191.16876101493835 seconds
Prediction Runtime: 2.427534341812134 seconds
Evaluation Runtime: 0.008106470108032227 seconds


After using hyper parameters, we see a couple of interesting observations. The first of which is that this process takes a long time to complete. Taking about 14 minutes to compute.

Within the best hyperparameters for modeling we see that

* Because n_estimators is 100 this means that at the best model, there will be 100 decision trees.
* min_samples_split tells us the minimum number of samples required to split into a node. The best is 2. Meaning a node is only split if it contains at least 2 samples
* min_samples_leaf is the minimum number of samples required to be at a leaf node. The best value is 1, indicating that a leaf node can have as few as 1 sample.
* max_depth is the maximum depth of the tree. The best value is None, which means there is no maximum depth limit, and the tree will grow until all leaves are pure or until all leaves contain a minimum number of samples.


With the best score being 0.9900522703553676 meaning that the best score that can be achieved within the model is about .99.  Meaning that the best Random Forest model acheived an average accuracy of 99% during cross validation within the search. Even with cross validation, this may infer that there is oversampling among the data after cross validation.

### Evaluating the Best Model Performance of the Random Classification

In [92]:
import time
from sklearn.metrics import classification_report, roc_auc_score

# Train set performance
start_time_train = time.time()
train_pred = best_model.predict(X_train)
train_auc = roc_auc_score(y_train, train_pred)
end_time_train = time.time()

print("Train Set Performance:")
print("AUC:", train_auc)
print(classification_report(y_train, train_pred))

# Test set performance
start_time_test = time.time()
test_pred = best_model.predict(X_val)
test_auc = roc_auc_score(y_val, test_pred)
end_time_test = time.time()

print("Test Set Performance:")
print("AUC:", test_auc)
print(classification_report(y_val, test_pred))

# Print the runtimes
print("Training Runtime:", end_time_train - start_time_train, "seconds")
print("Testing Runtime:", end_time_test - start_time_test, "seconds")


Train Set Performance:
AUC: 1.0
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    226132
           1       1.00      1.00      1.00     19876

    accuracy                           1.00    246008
   macro avg       1.00      1.00      1.00    246008
weighted avg       1.00      1.00      1.00    246008

Test Set Performance:
AUC: 0.507036696323935
              precision    recall  f1-score   support

           0       0.92      1.00      0.96     56554
           1       0.48      0.02      0.03      4949

    accuracy                           0.92     61503
   macro avg       0.70      0.51      0.49     61503
weighted avg       0.88      0.92      0.88     61503

Training Runtime: 9.18987774848938 seconds
Testing Runtime: 2.0379745960235596 seconds


Under the best parameters set for the Random Forest Modeling

The Train Set
* Has an AUC of 1.0. Meaning that the models predictive powers seem to be low because there is no diversity within the predictions. This could be a result of oversampling within the parameters.
* The precision, recall, and F1 score for the majority class of 0 (Those who do not default) are all 1.0. Very high meaning that the model performs well in identifying the majority class, perfection. This is a too high of a rating and could be overfitting.
* The precision, recall, and F1 score for the minority class of 1 (Those who do default) are at 1.0. Meaning that the model is a perfect classifier of those that are safe to lend to.
* This train set gives us a 100% accuracy rating. Meaning that this model at it's best can predict 100% of the outcomes for the borrower. This is not a good measurement against the train set because it is over fitting to the training data. 

The Test Set
* AUC of AUC: 0.507036696323935. This shows us that within the test set, the model does have a discriminatory power, but it is relatively low.
* The precision, recall, and F1 score are .92, 1.00, and .96 when the class of 0 (Those who are save borrowers). This means that the model predicts the class of 0 poorly as the data was overfitted in the process.
* The precision, recall, and F1 score are .48, 0.02, and 0.03 when the class of 1 (Those who are NOT save borrowers). This means that the model predicts the class of 1 poorly.
* We see that because of the macro average and the weighted average they are unbalanced between the classes. This may be why the model is better at predicting for class 0 than 1. Because of this the AUC would be a better measurement rather than using the accuracy %.
* The accuracy is 92% which again shows us that this model is accurate most often when predicting for the majority class of 0 , but not for the class of 1. This is a too high accuracy and the AUC remains a better score for prediction. \

### Features of the Best Random Forest Classification Model

In [93]:
import time
# Feature importance
start_time = time.time()
feature_importances = best_model.feature_importances_

# Create a DataFrame to store the feature importances
importance_df = pd.DataFrame({'Feature': X_train.columns, 'Importance': feature_importances})

# Sort the DataFrame by importance in descending order
importance_df = importance_df.sort_values('Importance', ascending=False)

# Display the top 10 most important features
print(importance_df.head(10))
end_time = time.time()

# Print the runtime
print("Runtime:", end_time - start_time, "seconds")


                   Feature  Importance
27            EXT_SOURCE_2  0.09877625
28            EXT_SOURCE_3  0.09503074
7               DAYS_BIRTH  0.05982140
10         DAYS_ID_PUBLISH  0.05542707
4              AMT_ANNUITY  0.05442446
9        DAYS_REGISTRATION  0.05414072
8            DAYS_EMPLOYED  0.05385393
33  DAYS_LAST_PHONE_CHANGE  0.05312109
0               SK_ID_CURR  0.05280328
3               AMT_CREDIT  0.04984712
Runtime: 0.11169624328613281 seconds


From this output, we learn that the top 10 most important features when seeing if someone is safe to lend money to are based on


* Normalized score from external data source
* Client's age in days at the time of application
* How many days before the application did client change the identity document
* Loan annuity
* How many days before the application did client change their registration
* How many days before the application the person started current employment
* Number of days since the client last changed their phone number
* ID of loan in our sample
* Credit amount of the loan

### Prediction on the Test Set

In [94]:
#Make predictions on the test set
test_predictions = best_model.predict(test_data)

In [95]:
#Create a submission DataFrame
submission = pd.DataFrame({
    "SK_ID_CURR": test_data["SK_ID_CURR"], 
    "TARGET": test_predictions
})

#Remove duplicate SK_ID_CURR values
submission = submission.drop_duplicates(subset="SK_ID_CURR", keep="first")

#Save the submission DataFrame to a CSV file
submission.to_csv("submission_randomforest3.csv", index=False)

After putting this model into Kaggle, we see that we get a Kaggle score of .50195. This matches our scores of the model both signaling that this model has been overfit and does not give as well predictions for the model as it should.

## Ensemble Model

### Creating the Ensemble Model with the RandomForest, Logistic Regression, and Gradient Boosting Models

In [96]:
import time
from sklearn.ensemble import VotingClassifier
from sklearn.model_selection import cross_val_score

# Initialize individual models
model1 = RandomForestClassifier()
model2 = LogisticRegression()
model3 = GradientBoostingClassifier()

# Evaluate individual models using cross-validation
start_time_cv = time.time()
model1_scores = cross_val_score(model1, X_train, y_train, cv=2, scoring='accuracy')
model2_scores = cross_val_score(model2, X_train, y_train, cv=2, scoring='accuracy')
model3_scores = cross_val_score(model3, X_train, y_train, cv=2, scoring='accuracy')
end_time_cv = time.time()

# Create the ensemble model
ensemble_model = VotingClassifier(estimators=[('model1', model1), ('model2', model2), ('model3', model3)])

# Evaluate ensemble model using cross-validation
start_time_ensemble_cv = time.time()
ensemble_scores = cross_val_score(ensemble_model, X_train, y_train, cv=5, scoring='accuracy')
end_time_ensemble_cv = time.time()

# Fit the ensemble model on the entire training set
start_time_fit = time.time()
ensemble_model.fit(X_train, y_train)
end_time_fit = time.time()

# Make predictions with the ensemble model
start_time_preds = time.time()
ensemble_preds = ensemble_model.predict(X_val)
end_time_preds = time.time()

# Calculate ensemble model accuracy
start_time_accuracy = time.time()
# Calculate accuracy for each individual model
model1_accuracy = np.mean(model1_scores)
model2_accuracy = np.mean(model2_scores)
model3_accuracy = np.mean(model3_scores)
ensemble_accuracy = accuracy_score(y_val, ensemble_preds)
end_time_accuracy = time.time()

# Print individual models and ensemble model performance
print("Individual Random Forrest Classifier Accuracy:", model1_accuracy)
print("Individual Logistic Regression Accuracy:", model2_accuracy)
print("Individual Gradient Boosting Classifier Accuracy:", model3_accuracy)
print("Ensemble Model Accuracy:", ensemble_accuracy)
print("Ensemble Model Cross-Validation Scores:", ensemble_scores)

# Print the runtimes
print("Cross-Validation Runtime:", end_time_cv - start_time_cv, "seconds")
print("Ensemble Cross-Validation Runtime:", end_time_ensemble_cv - start_time_ensemble_cv, "seconds")
print("Fit Runtime:", end_time_fit - start_time_fit, "seconds")
print("Prediction Runtime:", end_time_preds - start_time_preds, "seconds")
print("Accuracy Calculation Runtime:", end_time_accuracy - start_time_accuracy, "seconds")


Individual Random Forrest Classifier Accuracy: 0.9193278267373419
Individual Logistic Regression Accuracy: 0.9192018145751357
Individual Gradient Boosting Classifier Accuracy: 0.9192790478358427
Ensemble Model Accuracy: 0.9196299367510528
Ensemble Model Cross-Validation Scores: [0.9192919  0.91925125 0.91925125 0.91931058 0.91918863]
Cross-Validation Runtime: 275.05734634399414 seconds
Ensemble Cross-Validation Runtime: 1118.3136358261108 seconds
Fit Runtime: 297.7979905605316 seconds
Prediction Runtime: 2.6046464443206787 seconds
Accuracy Calculation Runtime: 0.007262468338012695 seconds


When we perform the ensemble model of all of the logistic Regression, Random Forest, and Gradient Boosting Models we can see that the Ensemble Model has an accuracy of 0.9196299367510528 or 91.96% accurate when predicting for the outcome of whether the customer will be a safe borrower or not.

When each of the models are combined, this gives us another issue of oversampling and over fitting to the model. This is partially because all models except for the Gradient Boost Model have been over fitted to the data.

### Up and Down Sampling

In [97]:
import time
from sklearn.metrics import accuracy_score
from sklearn.utils import resample

# Analyze class distribution in the training set
class_counts = y_train.value_counts()
print("Class Distribution (Before):")
print(class_counts)

# Upsample the minority class
df_majority = X_train[y_train == 0]
df_minority = X_train[y_train == 1]

df_minority_upsampled = resample(df_minority, replace=True, n_samples=len(df_majority), random_state=42)

X_train_upsampled = pd.concat([df_majority, df_minority_upsampled])
y_train_upsampled = pd.Series([0] * len(df_majority) + [1] * len(df_minority_upsampled))

# Downsample the majority class
df_majority_downsampled = resample(df_majority, replace=False, n_samples=len(df_minority), random_state=42)

X_train_downsampled = pd.concat([df_majority_downsampled, df_minority])
y_train_downsampled = pd.Series([0] * len(df_majority_downsampled) + [1] * len(df_minority))

# Train and evaluate models on upsampled and downsampled data
model = ensemble_model  # Initialize the Ensemble model created

# Train on upsampled data
start_time_upsampled = time.time()
model.fit(X_train_upsampled, y_train_upsampled)
end_time_upsampled = time.time()

# Make predictions on upsampled data
start_time_pred_upsampled = time.time()
y_pred_upsampled = model.predict(X_val)
end_time_pred_upsampled = time.time()

# Calculate accuracy on upsampled data
start_time_accuracy_upsampled = time.time()
accuracy_upsampled = accuracy_score(y_val, y_pred_upsampled)
end_time_accuracy_upsampled = time.time()

# Train on downsampled data
start_time_downsampled = time.time()
model.fit(X_train_downsampled, y_train_downsampled)
end_time_downsampled = time.time()

# Make predictions on downsampled data
start_time_pred_downsampled = time.time()
y_pred_downsampled = model.predict(X_val)
end_time_pred_downsampled = time.time()

# Calculate accuracy on downsampled data
start_time_accuracy_downsampled = time.time()
accuracy_downsampled = accuracy_score(y_val, y_pred_downsampled)
end_time_accuracy_downsampled = time.time()

# Compute predictions for original data
start_time_pred_original = time.time()
y_pred_original = model.predict(X_val)
end_time_pred_original = time.time()

# Calculate accuracy on original data
start_time_accuracy_original = time.time()
accuracy_original = accuracy_score(y_val, y_pred_original)
end_time_accuracy_original = time.time()

# Print model performance and runtimes
print("Accuracy (Original):", accuracy_original)
print("Accuracy (Upsampled):", accuracy_upsampled)
print("Accuracy (Downsampled):", accuracy_downsampled)

print("Upsampled Training Runtime:", end_time_upsampled - start_time_upsampled, "seconds")
print("Upsampled Prediction Runtime:", end_time_pred_upsampled - start_time_pred_upsampled, "seconds")
print("Upsampled Accuracy Calculation Runtime:", end_time_accuracy_upsampled - start_time_accuracy_upsampled, "seconds")

print("Downsampled Training Runtime:", end_time_downsampled - start_time_downsampled, "seconds")
print("Downsampled Prediction Runtime:", end_time_pred_downsampled - start_time_pred_downsampled, "seconds")
print("Downsampled Accuracy Calculation Runtime:", end_time_accuracy_downsampled - start_time_accuracy_downsampled, "seconds")

print("Original Prediction Runtime:", end_time_pred_original - start_time_pred_original, "seconds")
print("Original Accuracy Calculation Runtime:", end_time_accuracy_original - start_time_accuracy_original, "seconds")


Class Distribution (Before):
0    226132
1     19876
Name: TARGET, dtype: int64
Accuracy (Original): 0.6838203014487099
Accuracy (Upsampled): 0.7884493439344422
Accuracy (Downsampled): 0.6838203014487099
Upsampled Training Runtime: 454.61707830429077 seconds
Upsampled Prediction Runtime: 2.442786931991577 seconds
Upsampled Accuracy Calculation Runtime: 0.017435789108276367 seconds
Downsampled Training Runtime: 47.265679121017456 seconds
Downsampled Prediction Runtime: 2.351642370223999 seconds
Downsampled Accuracy Calculation Runtime: 0.004544496536254883 seconds
Original Prediction Runtime: 1.99517822265625 seconds
Original Accuracy Calculation Runtime: 0.008304834365844727 seconds


Within our Ensemble Model we see that the method that has the highest accuracy is when we use the Upsample dataset with an accuracy of 0.7884493439344422 or 78.84%. The original and downsampled dataset have an accuracy of 0.6838203014487099 or 68.38%. The Upsample data will now be used within our hyper parameters and evaluation of the model and features.These percentages show us that there is less overfitting happening because of the diversity within the percentages.

### Hyper Parameters

In [132]:
import time
from tqdm import tqdm
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

# Define the hyperparameters and their search space
param_grid = {
    'model1__n_estimators': [50, 100],
    'model1__max_depth': [None],
    'model1__min_samples_split': [2],
    'model1__min_samples_leaf': [1],
    'model2__C': [1.0],
    'model2__penalty': ['l2'],
    'model2__solver': ['liblinear']
}

# Reduce the number of cross-validation folds
cv = 2

# Split the training data into a smaller subset
X_train_subset, _, y_train_subset, _ = train_test_split(X_train_upsampled, y_train_upsampled, train_size=0.5, random_state=42)

# Initialize base estimators
model1 = RandomForestClassifier(n_estimators=50, random_state=42)  # Reduced number of estimators
model2 = LogisticRegression(random_state=42, max_iter=1000)  # Increased max_iter for faster convergence
model3 = GradientBoostingClassifier(random_state=42)

# Initialize ensemble model
ensemble_model = VotingClassifier(estimators=[('model1', model1),
                                              ('model2', model2),
                                              ('model3', model3)])

# Perform grid search with cross-validation
start_time_search = time.time()
grid_search = GridSearchCV(ensemble_model, param_grid, cv=cv, scoring='accuracy', n_jobs=-1)

# Wrap the grid search with tqdm and display progress and ETA
with tqdm(total=cv * len(grid_search.param_grid), ncols=80,
          bar_format='{l_bar}{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}, {rate_fmt}{postfix}], ETA: {remaining_s:.0f}s') as pbar:
    grid_search.fit(X_train_subset, y_train_subset)
    pbar.update(cv * len(grid_search.param_grid))  # Manually update progress bar to completion

end_time_search = time.time()

# Get the best hyperparameter values and model performance
best_params = grid_search.best_params_
best_score = grid_search.best_score_

# Train the model with the best hyperparameters
best_rf_params = {k.replace('model1__', ''): v for k, v in best_params.items() if k.startswith('model1__')}
best_lr_params = {k.replace('model2__', ''): v for k, v in best_params.items() if k.startswith('model2__')}
best_model = VotingClassifier(estimators=[('model1', RandomForestClassifier(**best_rf_params, random_state=42)),
                                          ('model2', LogisticRegression(**best_lr_params, random_state=42, max_iter=1000)),
                                          ('model3', GradientBoostingClassifier(**best_rf_params, random_state=42))])

# Wrap the fitting process with tqdm and display progress and ETA
with tqdm(total=len(X_train_upsampled), ncols=80,
          bar_format='{l_bar}{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}, {rate_fmt}{postfix}], ETA: {remaining_s:.0f}s') as pbar:
    best_model.fit(X_train_upsampled, y_train_upsampled)
    pbar.update(len(X_train_upsampled))  # Manually update progress bar to completion

# Make predictions using the best model
start_time_pred = time.time()
y_pred = best_model.predict(X_val)
end_time_pred = time.time()

# Evaluate the model performance
start_time_accuracy = time.time()
accuracy = accuracy_score(y_val, y_pred)
end_time_accuracy = time.time()

# Print the best hyperparameters and model performance
print("Best Hyperparameters:", best_params)
print("Best Score:", best_score)
print("Accuracy:", accuracy)

# Print the runtimes
print("Grid Search Runtime:", end_time_search - start_time_search, "seconds")
print("Fit Runtime:", end_time_pred - start_time_pred, "seconds")
print("Prediction Runtime:", end_time_pred - start_time_pred, "seconds")
print("Accuracy Calculation Runtime:", end_time_accuracy - start_time_accuracy, "seconds")


100%|█████████████████████████| 452264/452264 [54:00<00:00, 139.54it/s], ETA: 0s


Best Hyperparameters: {'model1__max_depth': None, 'model1__min_samples_leaf': 1, 'model1__min_samples_split': 2, 'model1__n_estimators': 100, 'model2__C': 1.0, 'model2__penalty': 'l2', 'model2__solver': 'liblinear'}
Best Score: 0.7919622167583535
Accuracy: 0.890493146675772
Grid Search Runtime: 459.12626457214355 seconds
Fit Runtime: 4.412505865097046 seconds
Prediction Runtime: 4.412505865097046 seconds
Accuracy Calculation Runtime: 0.021256446838378906 seconds


After using hyper parameters, we see a couple of interesting observations. The first of which is that this process takes a long time to complete. Taking about 10 minutes to compute.

Within the best hyperparameters for modeling we see that


* model1__max_depth: In our scenario, it is set to None, meaning that there is no maximum depth limit.

* model1__min_samples_leaf: Here, it is set to 1, meaning each leaf node should have at least one sample.

* model1__min_samples_split: It is set to 2, indicating that a node must have at least 2 samples to be split.  It may have more, but it must be a minimum of 2.

* model1__n_estimators: The best value found during the grid search is 100. This means that it takes about 100 decision trees to come up with the best parameter of the model.

* model2__C: This is used for the combination of the Logisitic Regtression within our model. The best value is 1.

* model2__penalty:  The best choice is l2, which refers to ridge regularization.This ridge regularization may have also been thrown off because of the parameters set to reduce computation time and capacity.

* model2__solver: The best choice is liblinear, which is suitable for small datasets. This means that this model is going to have the best results when performed on small or medium datasets, rather than a large one.



With the best score being 0.7919622167583535 meaning that the best score that can be achieved within the ensemble model is about .79. Meaning that the best Ensemble model acheived an average accuracy during cross validation within the search. This means that the model performs extremely well across the different data types.

With best accuracy of  0.890493146675772 meaning that in the best case scenario this model will be having the proportion of correct classifications with an unkown dataset of about .89 correct classifications. This means that at the best this model will still classify about .89 of the dataset correctly. While this may be an improvement from our other models, this is still alarming that the model needed to be less over fit to the data provided into the model.

### Evaluating the Best of the Ensemble Model

In [133]:
import time
from sklearn.metrics import classification_report, roc_auc_score

# Train set performance
start_time_train = time.time()
train_pred = best_model.predict(X_train)
train_auc = roc_auc_score(y_train, train_pred)
end_time_train = time.time()

# Print train set performance and runtime
print("Train Set Performance:")
print("AUC:", train_auc)
print(classification_report(y_train, train_pred))
print("Train Set Performance Runtime:", end_time_train - start_time_train, "seconds")

# Test set performance
start_time_test = time.time()
test_pred = best_model.predict(X_val)
test_auc = roc_auc_score(y_val, test_pred)
end_time_test = time.time()

# Print test set performance and runtime
print("Test Set Performance:")
print("AUC:", test_auc)
print(classification_report(y_val, test_pred))
print("Test Set Performance Runtime:", end_time_test - start_time_test, "seconds")


Train Set Performance:
AUC: 1.0
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    226132
           1       1.00      1.00      1.00     19876

    accuracy                           1.00    246008
   macro avg       1.00      1.00      1.00    246008
weighted avg       1.00      1.00      1.00    246008

Train Set Performance Runtime: 14.746994495391846 seconds
Test Set Performance:
AUC: 0.5319638928664842
              precision    recall  f1-score   support

           0       0.92      0.96      0.94     56554
           1       0.18      0.10      0.13      4949

    accuracy                           0.89     61503
   macro avg       0.55      0.53      0.54     61503
weighted avg       0.86      0.89      0.88     61503

Test Set Performance Runtime: 3.51649808883667 seconds


Under the best parameters set for the Ensemble Modeling

The Train Set
* Has an AUC of 1.0. Meaning that the models predictive powers seem to be a low discriminative power of the model. Indication that this Ensemble Model is over fitted to the data set. This is not a good measurement as it signals that it is just as comparable to random sampling.
* The precision, recall, and F1 score for the majority class of 0 (Those who do not default) are all 1.0. Very high meaning that the model performs well in identifying the majority class, perfection.
* The precision, recall, and F1 score for the minority class of 1 (Those who do default) are at 1.0. Meaning that the model is a perfect classifier of those that are safe to lend to.
* This train set gives us a 100% accuracy rating. This is a suspicious measurement as the model is potentially over fitting to the training set almost entirely, despite going through the process of cross validation.

The Test Set
* AUC of Test: 0.5319638928664842. This shows us that within the test set, the model does have a discrimatory power, but it is relevatively low. It does show us that although it may be low, against the test set there is a slight increase from being away from nearly random guessing. Not as high as it could be though.
* The precision, recall, and F1 score are  0.92, 0.96, and 0.94 when the class of 0 (Those who are save borrowers). This means that the model predicts the class of 0 not well.
* The precision, recall, and F1 score are  0.18, 0.10, and 0.13  when the class of 1 (Those who are NOT save borrowers). This means that the model predicts the class of 1 moderately. This is a disadvantage to our model because we are wanting it to be trained to better predict those that would be classified as 1 as knowing who will default on their loans.
* We see that because of the macro average and the weighted average they are unbalanced between the classes. 
* The accuracy is 0.89 or 89% which again shows us that although there are some improvements with this model from preventing over fitting, it still does have a tendency to over fit.

### Features of the Best Ensemble Model

In [137]:
# #Feature importance
# feature_importances = best_model.feature_importances_

# #Create a DataFrame to store the feature importances
# importance_df = pd.DataFrame({'Feature': X_train.columns, 'Importance': feature_importances})

# #Sort the DataFrame by importance in descending order
# importance_df = importance_df.sort_values('Importance', ascending=False)

# #Display the top 10 most important features
# print(importance_df.head(20))


# Calculate feature importances for each individual model
importance_df = pd.DataFrame({'Feature': X_train.columns})

for name, model in best_model.named_estimators_.items():
    if hasattr(model, 'feature_importances_'):
        importance_df[name] = model.feature_importances_
    elif hasattr(model, 'coef_'):
        importance_df[name] = model.coef_[0]
    else:
        importance_df[name] = 0

# Calculate the average importance across all individual models
importance_df['Average_Importance'] = importance_df.iloc[:, 1:].mean(axis=1)

# Sort the DataFrame by average importance in descending order
importance_df = importance_df.sort_values('Average_Importance', ascending=False)

# Display the top 20 most important features
print(importance_df.head(10))

                       Feature     model1      model2     model3  \
27                EXT_SOURCE_2 0.09877625 -0.00000771 0.11965889   
28                EXT_SOURCE_3 0.09503074 -0.00000725 0.12221178   
7                   DAYS_BIRTH 0.05982140  0.00000206 0.06640491   
10             DAYS_ID_PUBLISH 0.05542707  0.00003227 0.06239986   
9            DAYS_REGISTRATION 0.05414072  0.00001582 0.06051397   
0                   SK_ID_CURR 0.05280328  0.00000103 0.06084362   
8                DAYS_EMPLOYED 0.05385393 -0.00000093 0.05723017   
4                  AMT_ANNUITY 0.05442446  0.00001633 0.05342831   
33      DAYS_LAST_PHONE_CHANGE 0.05312109  0.00018206 0.05381501   
6   REGION_POPULATION_RELATIVE 0.04318285 -0.00000008 0.04957898   

    Average_Importance  
27          0.07280914  
28          0.07241176  
7           0.04207612  
10          0.03928640  
9           0.03822350  
0           0.03788264  
8           0.03702772  
4           0.03595637  
33          0.03570605  
6

From this output, we learn that the top 10 most important features from our ensemble model when seeing if someone is safe to lend money to are based on


* Normalized score from external data source
* Client's age in days at the time of application
* How many days before the application did client change the identity document
* How many days before the application did client change their registration
* ID of loan in our sample
* How many days before the application the person started current employment
* Loan annuity
* Number of days since the client last changed their phone number
* Normalized population of region where client lives

### Prediction on the Test Set

In [138]:
#Make predictions on the test set
test_predictions = best_model.predict(test_data)

In [139]:
#Create a submission DataFrame
submission = pd.DataFrame({
    "SK_ID_CURR": test_data["SK_ID_CURR"], 
    "TARGET": test_predictions
})

#Remove duplicate SK_ID_CURR values
submission = submission.drop_duplicates(subset="SK_ID_CURR", keep="first")

#Save the submission DataFrame to a CSV file
submission.to_csv("submission_ensemble2.csv", index=False)

When we put this dataset into Kaggle, we return a Kaggle score of .52188. This is an improvement from the Logistic and Random Forest, but it still is signaling toward potential over fitting and random guessing from the model because of the low Kaggle score. 

# Model Performances Summary

In [1]:
# Running this code that will display a summary table of all of the metrics that we have used to evaluate our models.

import pandas as pd
#!pip install openpyxl



# Set display options to show all rows and columns
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

# Read the Excel file
df = pd.read_excel(r"C:\Users\jusha\Documents\IS 6812 MSBA Capstone 1\Models Summary Modeling Assignment Excel.xlsx")

# Display the DataFrame as a table
df



Model Performance Evaluation (Based on Best)  \
0                                                 NaN   
1   Sample That gave Best Accuracy (Original, Upsa...   
2                                Best Hyperparameters   
3                                          Best Score   
4                                       Best Accuracy   
5                      Best Train Set Performance AUC   
6                 Best Train Set Performance Accuracy   
7   Best Train Set Precision, Recall, F1 Score for...   
8   Best Train Set Precision, Recall, F1 Score for...   
9                  Best Train Set Performance Runtime   
10                      Best Test Set Performance AUC   
11                 Best Test Set Performance Accuracy   
12                  Best Test Set Performance Runtime   
13  Best Test Set Precision, Recall, F1 Score for ...   
14  Best Test Set Precision, Recall, F1 Score for ...   
15  Top 10 Features in Relation to Customer Borrow...   
16                                       Kaggle Score   

                                           Unnamed: 1  \
0                             Gradient Boosting Model   
1                                           Upsampled   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   
5                                                 NaN   
6                                            0.744094   
7    0.17346568212080862, 0.5791068902808648, 0.26...   
8    0.826534317879192, 0.420893109719136, 0.73303...   
9                                                 NaN   
10                                           0.668819   
11                                                NaN   
12                                                NaN   
13                                                NaN   
14                                                NaN   
15  EXT_SOURCE_3, EXT_SOURCE_2 , AMT_REQ_CREDIT_BU...   
16                                           0.655604   

                                           Unnamed: 2  \
0                           Logistic Regression Model   
1                                            Original   
2   {'C': 0.1, 'penalty': 'l2', 'solver': 'libline...   
3                                            0.919202   
4                                            0.919516   
5                                            0.499998   
6                                                0.92   
7                            0.92      1.00      0.96   
8                            0.00      0.00      0.00   
9                         0.22581863403320312 seconds   
10                                           0.499991   
11                                               0.92   
12                        0.03416752815246582 seconds   
13                           0.92      1.00      0.96   
14                           0.00      0.00      0.00   
15  DAYS_LAST_PHONE_CHANGE, DAYS_ID_PUBLISH, DAYS_...   
16                                                0.5   

                                           Unnamed: 3  \
0                        Random Forest Classification   
1                                           Upsampled   
2   {'n_estimators': 100, 'min_samples_split': 2, ...   
3                                            0.990052   
4                                            0.919419   
5                                                   1   
6                                                   1   
7                           1.00      1.00      1.00    
8                            1.00      1.00      1.00   
9                            9.18987774848938 seconds   
10                                           0.507037   
11                                               0.92   
12                         2.0379745960235596 seconds   
13                           0.92      1.00      0.96   
14                           0.48      0.02      0.03   
15  EXT_SOURCE_2

Based on the AUC and Kaggle score, we see that the best model that has been created was the Gradient Boosting Model. 

The Gradient Boosting Model gives a Kaggle score of .656 and an AUC of .669.

These results show that the Gradient Boost Model created will provide a sufficient amount of accurate predictions, while the other models suffer from being over fitted. The other models created with lower AUC and lower Kaggle scores within the test set indicate that they are similar odds of simply guessing the outcome of the customer default status. The Gradient Boosting Model created will be the best one to use to determine which of the customers are going to default or if they are safe to approve a loan to. 

# Results Summary

After performing our Gradient Boost modeling processes we have discovered useful findings for Home Credit to consdier when knowing attributes about a client that is likely to default.

Attributes about a client that commonly infer a customer defaulting are if: 

* There are many enquiries to Credit Bureau 1 day before application. 

* The client provided work phone as their phone contact information.

* The client is located in a low rated regions. 

* The loan is used for high price of goods that are bought by the client.

* There are many enquiries to Credit Bureau 1 hour before application.

* The permanent address provided by the client is not the same as the contact address provided by the client.

By knowing each of these, this solves our business problem by knowing which demographics or audiences who may not be the most safe to loan to. By now knowing each of these attributes and using the model, this will help to lower risk of lending to someone who will not repay their loan, thus increasing the cost. By only approving loans to customers who are predicted to repay their loan, Home Credit can increase their revenue. By increasing revenue and lowering costs, in the end profits will be increased.

# Justin Hamilton Individual Contribution Summary

***Justin Hamilton:*** I first ran a trial model for the Gradient Boost, Logistic Regression, Random Forest Classifier, and Ensemble Model. I created this in my own document and then shared it within the group. After seeing the rest of the groups efforts, I then compiled all of the codes and models into this document and organized the files. I wrote the interpretations within this notebook and created the Logistic, Random Forest, and Ensemble Models. I wrote up the summary rough draft, made the interpretations throughout the analysis, and compiled this whole document neatly for submission. I also wrote the model hyper parameters, evaluations, and feature engineering for each of the models to come to a conclusion. Wrote the results summary of the modeling.
